API resource Record creates Request objects in v2

xliiv opened the pull request:
API is now versioned (default version is 'v1').
In version 'v2' resource Record creates always Requests (RecordRequest for
creation and update, DeleteRequest for deletion)
Beside that Record endpoint also checks if requested changes can be
autoaccepted, which instantly changes Records in powerdns table.

szok commented on powerdns/serializers.py:
```diff
@@ -38,6 +39,24 @@ class Meta:
         read_only_fields = ('notified_serial',)
 
 
+class RecordRequestSerializer(HyperlinkedModelSerializer):
```
use OwnedSerialiazer

szok commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+            serializer_class = RecordRequestSerializer
+        else:
+            serializer_class = RecordSerializer
+        print(serializer_class)
```
remove :)

szok commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+        # TODO:: this code needs some love, any idea?
+        auto_accept = (
+            self.request.user.is_superuser or
+            serializer.instance.domain.unrestricted == True or
```
delete "== true"

szok commented on powerdns/views.py:
```diff
@@ -65,6 +68,13 @@ class DomainViewSet(OwnerViewSet):
     permission_classes = (DomainPermission,)
 
 
+class RecordRequestsViewSet(ReadOnlyModelViewSet):
```
Add new filter by "owner" to Record and Domain

xliiv replied:
done

xliiv commented:
merge in tests

mkurek commented on dnsaas/urls.py:
```diff
@@ -34,11 +36,14 @@
 router.register(r'domains', DomainViewSet)
 router.register(r'records', RecordViewSet)
+#TODO:: better name?
+router.register(r'gui-records', GUIRecordViewSet, base_name='gui')
```
could we expose it as `records` in v2 version of the API?

mkurek commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+
+        # TODO:: this code needs some love, any idea?
```
maybe move it to the model?

mkurek commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+            request.user.id in serializer.instance.domain.authorisations.values_list('authorised', flat=True) # noqa
+        )
+        if auto_accept:
+            record = serializer.instance.accept()
```
couldn't `accept` assign record?

mkurek commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+    def update(self, request, *args, **kwargs):
+        partial = kwargs.pop('partial', False)
+        instance = get_object_or_404(
+            RecordRequest, record__pk=self.get_object().pk)
```
are you sure it'd work? as far as i understand it, you're fetching here `RecordRequest` matching to `Record` pointed by current pk (from url). What if there is no such Request? What if there is more than one such Request? 
IMO `update` on viewset should:
- get `Record` instance
- create `RecordRequest` (without saving) based on data from `Record`
- then perform regular update (or even create) flow (like calling `get_serializer` and `is_valid`)
- instead of calling `perform_update`, we want to call `perform_create` to create new instance of `RecordRequest`

Most of this method would be the same as in `create`.

mkurek commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
 
 
+
+class GUIRecordViewSet(OwnerViewSet):
```
what about `DELETE` method?

mkurek commented on powerdns/views.py:
```diff
@@ -92,6 +102,94 @@ def get_queryset(self):
+            headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=code, headers=headers)
+
+    def get_queryset(self):
```
this is the same as in the `RecordViewSet`. Maybe we could move it somewhere to common place? Maybe some Record manager? Then calling `Record.objects.filter(ip='1.2.3.4')` will make filter like here. Other thing is, that this logic (filtering) should be moved to some filter backend (maybe sth like http://www.django-rest-framework.org/api-guide/filtering/#specifying-a-filterset ?)

mkurek commented:
make sure that there will be test covering the full flow of the Record (creating, updating, deleting), including multiple users and partial updates :)

szok commented:
:ok:

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,120 @@ class DomainViewSet(OwnerViewSet):
+        return obj
+
+    def get_serializer_class(self):
+        if self.request.version == 'v1':
```
create, update and delete also should work that way, am i right?

xliiv replied:
yes, `delete` is up to implement..

xliiv replied:
done

Commit:
API resource Record creates Request objects in v2

API is now versioned (default version is 'v1').
In version 'v2' resource Record creates always Requests (RecordRequest for
creation and update, DeleteRequest for deletion)
Beside that Record endpoint also checks if requested changes can be
autoaccepted, which instantly changes Records in powerdns table.

mkurek commented on powerdns/models/requests.py:
```diff
@@ -242,6 +262,8 @@ class RecordRequest(ChangeCreateRequest, RecordLike):
     record = models.ForeignKey(
         Record,
+        # used for history purpose
+        on_delete=models.DO_NOTHING,
```
i'd consider here https://docs.djangoproject.com/en/dev/ref/models/fields/#django.db.models.ForeignKey.db_constraint too (maybe it's even better than DO_NOTHING (i'm not sure how this exactly works in mysql))

mkurek commented on powerdns/models/powerdns.py:
```diff
@@ -390,6 +390,53 @@ class Meta:
         verbose_name = _("record")
         verbose_name_plural = _("records")
 
+    def request_factory(operation):
```
this looks like class-level method, but it's regular function, isn't it? if so, then why it's defined in this class? :)

mkurek replied:
and some docstring will be useful here

mkurek replied:
and since this is returning links in admin context, why this is attached directly to model?

xliiv replied:
This copy-paste from: https://github.com/allegro/django-powerdns-dnssec/blob/develop/powerdns/models/powerdns.py#L126
With only one small change:
```
 +                ) and
 +                can_edit(get_current_user(), self) or
 +                can_delete(get_current_user(), self)
```
Why gonna get rid of it all when admin become simple CRUD.
No point of changing it, IMO :)

mkurek replied:
ok

mkurek commented on powerdns/models/requests.py:
```diff
@@ -319,5 +341,9 @@ def get_object(self):
         else:
             return Record(domain=self.domain, owner=self.owner)
 
+    def accept_and_assign(self):
```
assign who/what?

mkurek commented on powerdns/tests/test_api.py:
```diff
@@ -62,21 +68,195 @@ def test_record_filters_by_ip_and_type(self):
         mvs.request = request
         self.assertEqual(len(mvs.get_queryset()), 8)
 
-    def test_no_access_to_not_owned_domain(self):
```
shouldnt this still work in v1 api?

xliiv replied:
this is duplicated, so i removed it

mkurek replied:
ok

mkurek commented on powerdns/tests/test_api.py:
```diff
@@ -62,21 +68,195 @@ def test_record_filters_by_ip_and_type(self):
 
-        url = reverse('domain-detail', args=(domain_id,))
-        response = self.client.get(url, format='json')
+    def test_record_is_not_created_when_domain_is_other(self):
```
is other? what do you mean by that?

xliiv replied:
change it to
`test_record_is_not_created_when_domain_is_not_owned` 
?

mkurek replied:
yes

mkurek commented on powerdns/tests/test_api.py:
```diff
@@ -62,21 +68,195 @@ def test_record_filters_by_ip_and_type(self):
+            format='json',
+            **{'HTTP_ACCEPT': 'application/json; version=v2'}
+        )
         self.assertEqual(response.status_code, status.HTTP_200_OK)
```
when there is already created request for some record, shouldn't we reject next request(s), until the first one will be finished?

xliiv replied:
done

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,149 @@ class DomainViewSet(OwnerViewSet):
+                serializer.initial_data.keys()
+            )
+        ]
+        record_request.copy_records_data(data_to_copy)
```
maybe we could use https://docs.djangoproject.com/en/1.9/ref/models/instances/#specifying-which-fields-to-save here?

xliiv replied:
This `copy_records_data` copy data with a bit logic, so I don't see use of this here.

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,149 @@ class DomainViewSet(OwnerViewSet):
+        ]
+        record_request.copy_records_data(data_to_copy)
+        record_request.save()
+        auto_accept = (
```
this is almost a duplicate of `auto_accept` in create. could we move it to some function?

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,149 @@ class DomainViewSet(OwnerViewSet):
+        if auto_accept:
+            serializer.save()
+            record_request.record = serializer.instance
+            record_request.accept()
```
why `accept`, not `accept_and_assign`?

xliiv replied:
assign is done here
`record_request.record = serializer.instance`

mkurek replied:
is the reversed order of "events" here relevant?

xliiv replied:
`record_request.accept()` writes data from `record_request` to `record_request.record`.
Perhaps simply `record_request.state = RequestStates.ACCEPTED`
would be enough (instead of record_request.accept())

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,149 @@ class DomainViewSet(OwnerViewSet):
+
+    def destroy(self, request, *args, **kwargs):
+        if self.request.version == 'v1':
+            return super().create(request, *args, **kwargs)
```
create? are you sure? :)

xliiv replied:
Of course that this is bad ;)

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +70,149 @@ class DomainViewSet(OwnerViewSet):
+            owner=self.request.user, target=instance,
+        )
+        delete_request.save()
+        auto_accept = (
```
ditto

xliiv replied:
?

Commit:
Merge branch 'develop' into new-api

Conflicts:
	powerdns/tests/test_api.py
	powerdns/tests/utils.py

Commit:
flake8

Commit:
wip

Commit:
db_constraint

Commit:
reject updates when opened Request already created on Record

Commit:
review fixes

Commit:
auto_accept in create moved to domain model

Commit:
auto_accept moved to model on record.update()

Commit:
auto_accept uses models method in record.destroy()

Commit:
added opened-requests check to delete

mkurek commented on powerdns/tests/test_api.py:
```diff
@@ -76,21 +81,236 @@ def test_record_filters_by_ip_and_type(self):
+            format='json',
+            **{'HTTP_ACCEPT': 'application/json; version=v2'}
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
are you sure this should return 200 (and generally, work)? As discussed above, when record request already exists, next one should be rejected. Isn't it the case of this test? edit: is superuser playing some role here?

xliiv replied:
This is a bit redundant with: `test_dont_reject_update_when_already_exists_but_superuser`
I'll clean it.

xliiv replied:
done

mkurek commented on powerdns/tests/test_api.py:
```diff
@@ -76,21 +81,236 @@ def test_record_filters_by_ip_and_type(self):
+            DeleteRequest.objects.get(target_id=record_request.record_id)
+        )
+
+    def test_delete_record_when_cant_auto_accept(self):
```
make some additional test for deletion when another request is pending.

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +71,174 @@ class DomainViewSet(OwnerViewSet):
+
+        if (
+            serializer.instance.opened_requests and
+            not request.user.is_superuser
```
i'm not sure that fact that the user is superuser has any meaning here

xliiv replied:
This is how it works now. It could be easily changed in future.

mkurek commented on powerdns/views.py:
```diff
@@ -61,17 +71,174 @@ class DomainViewSet(OwnerViewSet):
+                        'id', flat=True
+                    )
+                 },
+                # TODO:: what return code?
```
412?

Commit:
review fixes

Commit:
Merge branch 'develop' into new-api

Conflicts:
	dnsaas/urls.py

Commit:
migrations

mkurek commented:
:ok:

Commit:
better test names

szok commented:
:ok:

Commit:
rm migration 22

Commit:
Merge branch 'develop' into new-api

Commit:
migrations

Commit:
Merge branch 'develop' into new-api

Commit:
migrations again, yawn
